src/uri.jl

```diff
@@ -145,13 +145,13 @@ function Base.isvalid(uri::URI)
 end
 
 lower(c::UInt8) = c | 0x20
-@inline shouldencode(c) = c < 0x1f || c > 0x7f || @anyeq(c,
-                            UInt8(';'), UInt8('/'), UInt8('?'), UInt8(':'),
-                            UInt8('@'), UInt8('='), UInt8('&'), UInt8(' '),
-                            UInt8('"'), UInt8('<'), UInt8('>'), UInt8('#'),
-                            UInt8('%'), UInt8('{'), UInt8('}'), UInt8('|'),
-                            UInt8('\\'), UInt8('^'), UInt8('~'), UInt8('['),
-                            UInt8(']'), UInt8('`'))
+# RFC3986 Unreserved Characters (and '~' Unsafe per RFC1738).
+@inline shouldencode(c) = !((c >= UInt8('A') && c <= UInt8('Z'))
+                         || (c >= UInt8('a') && c <= UInt8('z'))
+                         || (c >= UInt8('0') && c <= UInt8('9'))
+                         || c == UInt8('-')
+                         || c == UInt8('.')
+                         || c == UInt8('_'))
```

Comment:
Looking at this again, maybe this would be simpler:

```julia
shouldencode(c::Char) = !((isascii(c) && isalnum(c)) || c in "-._")
```

or perhaps:

```julia
const urlsafe = [(isascii(c) && isalnum(c)) || c in "-._" for c in Vector{Char}(0x00:0xFF)]
shouldencode(c) = !urlsafe[c+1]
```